Std/Data/BitVec/Basic.lean

```diff
@@ -0,0 +1,200 @@
+  { val := x.val % y.val }
+/--
+Division for bit vectors.
+Note that `div x 0 = allOnes n` as specified by the SMT-Lib standard.
+See the spec at http://smtlib.cs.uiowa.edu/theories-FixedSizeBitVectors.shtml.
+This is also consistent with a simple division circuit.
+-/
+protected def div (x y : BitVec n) : BitVec n :=
+  if y = BitVec.zero n then
+    BitVec.allOnes n
+  else
+    { val := x.val / y.val }
+/--
+Division for bit vectors using the Lean convention where division by zero returns zero.
+-/
+protected def div' (x y : BitVec n) : BitVec n :=
+  { val := x.val / y.val }
+/-- Less than for bit vectors. -/
+protected def lt (x y : BitVec n) : Bool :=
+  x.val < y.val
+/-- Less than or equal to for bit vectors. -/
+protected def le (x y : BitVec n) : Bool :=
+  x.val ≤ y.val
+
+instance : Add (BitVec n) := ⟨BitVec.add⟩
+instance : Sub (BitVec n) := ⟨BitVec.sub⟩
+instance : Mul (BitVec n) := ⟨BitVec.mul⟩
+instance : Mod (BitVec n) := ⟨BitVec.mod⟩
+instance : Div (BitVec n) := ⟨BitVec.div⟩
+instance : LT (BitVec n)  := ⟨fun x y => BitVec.lt x y⟩
+instance : LE (BitVec n)  := ⟨fun x y => BitVec.le x y⟩
+
+/-- Bitwise and for bit vectors. -/
+protected def and (x y : BitVec n) : BitVec n :=
+  { val := x.val &&& y.val }
+/-- Bitwise or for bit vectors. -/
+protected def or (x y : BitVec n) : BitVec n :=
+  { val := x.val ||| y.val }
+/-- Bitwise xor for bit vectors. -/
+protected def xor (x y : BitVec n) : BitVec n :=
+  { val := x.val ^^^ y.val }
+/-- Complement for bit vectors. -/
+protected def complement (x : BitVec n) : BitVec n :=
+  0 - (x + .ofNat n 1)
+/-- Shift left for bit vectors. -/
+protected def shiftLeft (a : BitVec n) (s : Nat) : BitVec n :=
+  .ofNat n (a.toNat <<< s)
+/-- Shift right for bit vectors. -/
+protected def shiftRight (a : BitVec n) (s : Nat) : BitVec n :=
+  .ofNat n (a.toNat >>> s)
+
+instance : Complement (BitVec w) := ⟨BitVec.complement⟩
+instance : AndOp (BitVec w) := ⟨BitVec.and⟩
+instance : OrOp (BitVec w) := ⟨BitVec.or⟩
+instance : Xor (BitVec w) := ⟨BitVec.xor⟩
+instance : HShiftLeft (BitVec w) Nat (BitVec w) := ⟨BitVec.shiftLeft⟩
+instance : HShiftRight (BitVec w) Nat (BitVec w) := ⟨BitVec.shiftRight⟩
+
+/-- Rotate left. -/
+def rotateLeft (x : BitVec w) (n : Nat) : BitVec w :=
+  x <<< n ||| x >>> (w - n)
+
+/-- Rotate right. -/
+def rotateRight (x : BitVec w) (n : Nat) : BitVec w :=
+  x >>> n ||| x <<< (w - n)
+
+/-- Append. -/
+def append (msbs : BitVec n) (lsbs : BitVec m) : BitVec (n+m) :=
+  .ofNat (n+m) (msbs.toNat <<< m + lsbs.toNat)
+
+instance : HAppend (BitVec w) (BitVec v) (BitVec (w+v)) := ⟨BitVec.append⟩
+
+/--
+Extraction of bits `i` down to `j` from a bit vector of size `n` to yield a
+new bitvector of size `i - j + 1`
+-/
+def extract (i j : Nat) (a : BitVec n) : BitVec (i - j + 1) :=
+  BitVec.ofNat _ (a.toNat >>> j)
+
+/--
+`repeat_ i x` means concatenate `i` copies of `x`.
+Recall that `repeat` is a keyword in Lean.
+-/
+def replicate : (i : Nat) → BitVec w → BitVec (w*i)
```

Comment:
```suggestion
-- TODO: write this using multiplication
def replicate : (i : Nat) → BitVec w → BitVec (w*i)
```
This should be equivalent to `x * (2^(w*i) - 1) / (2^w - 1)`, although there is some nontrivial proof to show it is in range and some even more nontrivial proof to show it replicates the bits in the same way as this linear time solution.

PS, while I have you here: `1 <<< n` on Nat does not have a builtin kernel implementation, which throws off the asymptotics of some algorithms. (I think you just get `2^n` implemented as `n` multiplications.) What do you think about adding bitwise ops on Nat to the list of accelerated kernel ops?

Reply:
I love these suggestions. It will help for sure people working on software and hardware verification.